flink-ml-lib/src/main/java/org/apache/flink/ml/common/glm/LocalTrainer.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.common.glm;
+
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeinfo.PrimitiveArrayTypeInfo;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.iteration.operator.OperatorStateUtils;
+import org.apache.flink.ml.linalg.DenseVector;
+import org.apache.flink.ml.regression.linearregression.LinearRegression;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+
+import java.util.Iterator;
+
+/**
+ * A local trainer is a trainer that uses a batch of training data to compute a update of the model
+ * locally.
+ *
+ * @param <T> Class type of training data.
+ */
+public abstract class LocalTrainer<T> implements CheckpointedFunction {
```

Comment:
For `LocalTrainer` and its subclasses, how do you like the following changes?
- move `CacheDataAndDoTrain` to an independent class
- merge `CacheDataAndDoTrain`, `LocalTrainer` and `WithRegularization`
- remove methods like `getReg`, as now they are only used internally
- change the API and implementation of `trainOnBatchData`. I have the sense that each time only one data, instead of a batch of data, is enough. I think the API of `org.apache.flink.api.common.functions.AggregateFunction` is a good reference.
- rename the merged class. A name like `LocalTrainer` might be too general to be associated with linear algorithms.

Reply:
> move CacheDataAndDoTrain to an independent class
It is an dependent class in the old PR. Do you mean to merge it with LocalTrainer and WithRegularization?

> merge LocalTrainer and WithRegularization and remove methods like getReg, as now they are only used internally
It seems hard to merge these two class, because the logic of `WithRegularization` is supposed to be used in `LocalTrainer#updateModel`. I have renamed `WithRegularization` as `RegularizationUtils` and removed methods like `withReg`. What do you think?

> rename the merged class. A name like LocalTrainer might be too general to be associated with linear algorithms.
I have renamed the class to `LocalLinearTrainer`. I still think it is not a very good name. We could probably discuss more on the naming.

> merge CacheDataAndDoTrain, LocalTrainer and WithRegularization.
I did not merge `CacheDataAndDoTrain` with the other two for now for the following reasons:
- `CacheDataAndDoTrain` is a more a infra and involves distributed concepts, like two input operators.
- `LocalTrainer` is a more friendly and clean concept for machine learning users, since it only involves local operations.

> change the API and implementation of trainOnBatchData. I have the sense that each time only one data, instead of a batch of data, is enough. I think the API of org.apache.flink.api.common.functions.AggregateFunction is a good reference.
I did not do the change for the following reason:
- mini-batch training is a common concept for machine learning.
- Users may want to do operations before/after mini-batch training.